doc/mgmt/SONiC_HAMD_RBAC_HLD.md

```diff
@@ -0,0 +1,1289 @@
+This attribute has been implemented in the pam_radius and libnss-radius modules as a part of this feature in SONiC.
+
+Hereafter, remote users can be assigned with groupnames, which are configured in the Radius server as follows:
+
+<user name>    Cleartext-Password := "force10"
+SONiC-User-Group := ["secadmin" | "netadmin" | "admin" | "operator"]
+
+The newly added groups (netadmin and secadmin) have been mapped to MPL=14 and MPL=13 respectively.
+
+2. Management-Privilege-Level
+The Management-Privilege-Level VSA (PW_MANAGEMENT_PRIVILEGE_LEVEL) has been added as a part of the recent RFC-5607.
+But it has been already implemented in SONiC. MPL = 1 to 15 can be used for the remote users.
+
+<user name>    Cleartext-Password := "force10"
+Management-Privilege-Level = [1 to 15]
+
+- Management-Privilege-Level = 15 => admin
+	- Changed the supplementary groups : from "admin,sudo,docker" to "admin,adm,redis,sudo,docker"
+
+- Management-Privilege-Level = 1 => operator
+	- Changed the supplementary groups : from "docker" to "operator,docker"
+
+- Management-Privilege-Level = 14 => netadmin
+	- Added supplementary groups : "netadmin,redis,docker"
+
+- Management-Privilege-Level = 13 => secadmin
+	- Added supplementary groups : "secadmin,redis,docker"
+
+1 <= MPL <= 12 has been assigned with operator role by default.
+
+## 2.2 Tacacs functionality
+
+Tacacs+ supports two main attributes: 'group-name' and 'priv-level'
+
+group-name has been implemented as a part of this feature.
+priv-level has been enhanced to support the two new roles : secadmin and netadmin.
+
+1. group-name
+Remote users can be assigned with groupnames, which are configured in the Radius server as follows:
+
+user = <username> {
+default service = permit
+pap = cleartext dellforce10
+service = exec {
+group-name = [admin|operator|secadmin|netadmin]                 
+}
+}
+
+The newly added groups (netadmin and secadmin) have been mapped to MPL=14 and MPL=13 respectively.
+
+2. priv-lvl
+Users with priv-lvl attribute can be created as follows :
+
+user = <username> {
+default service = permit
+pap = cleartext dellforce10
+service = exec {
+priv-lvl = [1 to 15]
+}
+}
+
+User priv-lvl = 15 => admin
+	- Changed the supplementary groups : from "sudo,docker" to "admin,adm,redis,sudo,docker"
+
+User priv-lvl = 1 => operator
+	- Changed the supplementary groups : from "users" to "operator,docker"
+	- Changed the default login shell : from "/bin/bash" to "/usr/bin/sonic-launch-shell"
+
+User priv-lvl = 14 => netadmin
+	- Added supplementary groups : "netadmin,redis,docker"
+
+User priv-lvl = 13 => secadmin
+	- Added supplementary groups : "secadmin,redis,docker"
+
+1 <= priv-lvl<= 12 has been assigned with operator role by default.
+
+# 3 Design
+## 3.1 HAM Design Overview
+
+The following diagram captures the current implementation of the command "username <username> password <password> role <role>". There will not be any changes in the following design.
```

Comment:
So HAMd will listen to notifications from CONFIG_DB for user role modifications, as well as DBUS calls from the RESTCONF server (specifically translib library) for user password modifications ? So, to confirm, the user passwords will never be stored in the redis.

Reply:
Yes, user passwords will not be stored in any of the redis DBs.

Yes, HAMd will listen to config_DB notifications & DBUS calls from RESTCONF server for both password and role modifications.

1. The password modification will result only in issuing a usermod to the linux kernel.
2. The role modification will result in executing a usermod to the linux kernel as well as updating the user role in the STATE_DB.

(The primary usecase for HAMd listening to config_DB notifications is to address the save & reload case issue. The user info from the state_DB is lost after a save & reload. To restore the data back in STATE_DB, HAMd subscribes to config_DB.

Since the config_DB notification does not contain the password of the user, we cant use that notification for any other purpose. The user name to user role mapping is alone is stored in the STATE_DB for RBAC feature using the config_DB notifications after a save&reload / upgrade scenario)